File: bot-sdk/go/bot.go
// Package botsdk provides a Go SDK for building Cats Company bots over WebSocket.
package botsdk

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"sync"
	"sync/atomic"
	"time"

	"github.com/gorilla/websocket"
)

// MessageHandler processes incoming messages.
type MessageHandler func(ctx *Context)

// Context provides message context to handlers.
type Context struct {
	Topic   string
	From    string
	Content string
	SeqID   int
	bot     *Bot
}

// Reply sends a text reply to the same topic.
func (c *Context) Reply(text string) error {
	return c.bot.SendMessage(c.Topic, text)
}

// ReplyWithTyping sends a typing indicator, waits briefly, then replies.
func (c *Context) ReplyWithTyping(text string) error {
	c.bot.SendTypingIndicator(c.Topic)
	time.Sleep(500 * time.Millisecond)
	return c.bot.SendMessage(c.Topic, text)
}

// Bot represents a connected bot instance using WebSocket.
type Bot struct {
	serverURL string
	apiKey    string
	conn      *websocket.Conn
	handlers  map[string]MessageHandler
	onReady   func()
	sendCh    chan []byte
	done      chan struct{}
	closeOnce sync.Once
	msgID     atomic.Int64
}

// New creates a new Bot instance.
// serverURL is the WebSocket endpoint, e.g. "ws://localhost:6061/v0/channels".
// apiKey is the bot's API key (format: cc_{hex_uid}_{random}).
func New(serverURL, apiKey string) *Bot {
	return &Bot{
		serverURL: serverURL,
		apiKey:    apiKey,
		handlers:  make(map[string]MessageHandler),
		sendCh:    make(chan []byte, 64),
		done:      make(chan struct{}),
	}
}

// OnMessage registers a handler for incoming data messages.
func (b *Bot) OnMessage(handler MessageHandler) {
	b.handlers["message"] = handler
}

// OnReady registers a callback for when the bot has completed the handshake.
func (b *Bot) OnReady(handler func()) {
	b.onReady = handler
}

// nextID returns a monotonically increasing message ID string.
func (b *Bot) nextID() string {
	return fmt.Sprintf("%d", b.msgID.Add(1))
}

// Connect establishes the WebSocket connection and performs the handshake.
func (b *Bot) Connect() error {
	url := fmt.Sprintf("%s?api_key=%s", b.serverURL, b.apiKey)
	conn, _, err := websocket.DefaultDialer.Dial(url, nil)
	if err != nil {
		return fmt.Errorf("ws dial: %w", err)
	}
	b.conn = conn

	// Send handshake
	hi := map[string]interface{}{
		"hi": map[string]interface{}{
			"id":  b.nextID(),
			"ver": "0.1.0",
		},
	}
	data, _ := json.Marshal(hi)
	if err := b.conn.WriteMessage(websocket.TextMessage, data); err != nil {
		b.conn.Close()
		return fmt.Errorf("handshake write: %w", err)
	}

	// Wait for handshake response
	_, msg, err := b.conn.ReadMessage()
	if err != nil {
		b.conn.Close()
		return fmt.Errorf("handshake read: %w", err)
	}

	var resp serverMessage
	if err := json.Unmarshal(msg, &resp); err != nil {
		b.conn.Close()
		return fmt.Errorf("handshake parse: %w", err)
	}
	if resp.Ctrl == nil || resp.Ctrl.Code != 200 {
		b.conn.Close()
		return fmt.Errorf("handshake rejected: %s", string(msg))
	}

	log.Printf("bot connected to %s", b.serverURL)

	if b.onReady != nil {
		b.onReady()
	}
	return nil
}

// Close shuts down the bot connection.
func (b *Bot) Close() {
	b.closeOnce.Do(func() {
		close(b.done)
		if b.conn != nil {
			b.conn.WriteMessage(
				websocket.CloseMessage,
				websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""),
			)
			b.conn.Close()
		}
	})
}

// SendMessage sends a text message to a topic.
func (b *Bot) SendMessage(topic, text string) error {
	msg := map[string]interface{}{
		"pub": map[string]interface{}{
			"id":      b.nextID(),
			"topic":   topic,
			"content": text,
		},
	}
	data, err := json.Marshal(msg)
	if err != nil {
		return err
	}
	select {
	case b.sendCh <- data:
		return nil
	case <-b.done:
		return fmt.Errorf("bot closed")
	}
}

// SendTypingIndicator sends a typing notification to a topic.
func (b *Bot) SendTypingIndicator(topic string) {
	msg := map[string]interface{}{
		"note": map[string]interface{}{
			"topic": topic,
			"what":  "kp",
		},
	}
	data, _ := json.Marshal(msg)
	select {
	case b.sendCh <- data:
	case <-b.done:
	}
}

// Run starts the bot's message loop (blocking). Cancel the context to stop.
func (b *Bot) Run(ctx context.Context) error {
	if err := b.Connect(); err != nil {
		return err
	}
	defer b.Close()

	// Writer goroutine
	go b.writePump()

	// Reader loop (blocking)
	return b.readPump(ctx)
}

// writePump sends queued messages to the WebSocket connection.
func (b *Bot) writePump() {
	for {
		select {
		case data := <-b.sendCh:
			if err := b.conn.WriteMessage(websocket.TextMessage, data); err != nil {
				log.Printf("bot write error: %v", err)
				return
			}
		case <-b.done:
			return
		}
	}
}

// readPump reads messages from the WebSocket and dispatches to handlers.
func (b *Bot) readPump(ctx context.Context) error {
	readErr := make(chan error, 1)

	go func() {
		for {
			_, raw, err := b.conn.ReadMessage()
			if err != nil {
				readErr <- err
				return
			}

			var msg serverMessage
			if err := json.Unmarshal(raw, &msg); err != nil {
				log.Printf("bot unmarshal error: %v", err)
				continue
			}

			if msg.Data != nil {
				handler, ok := b.handlers["message"]
				if ok {
					handler(&Context{
						Topic:   msg.Data.Topic,
						From:    msg.Data.From,
						Content: fmt.Sprintf("%v", msg.Data.Content),
						SeqID:   msg.Data.SeqID,
						bot:     b,
					})
				}
			}
		}
	}()

	select {
	case <-ctx.Done():
		return ctx.Err()
	case err := <-readErr:
		return fmt.Errorf("read pump: %w", err)
	case <-b.done:
		return nil
	}
}

// serverMessage mirrors the server-to-client envelope (subset needed by SDK).
type serverMessage struct {
	Ctrl *msgCtrl `json:"ctrl,omitempty"`
	Data *msgData `json:"data,omitempty"`
}

type msgCtrl struct {
	ID   string `json:"id,omitempty"`
	Code int    `json:"code"`
	Text string `json:"text,omitempty"`
}

type msgData struct {
	Topic   string      `json:"topic"`
	From    string      `json:"from,omitempty"`
	SeqID   int         `json:"seq"`
	Content interface{} `json:"content"`
}

File: bot-sdk/python/openchat_bot/bot.py
"""Cats Company Python Bot SDK - WebSocket-based bot framework."""

import json
import logging
import threading
import time
from typing import Callable, Optional

import websocket

logger = logging.getLogger("catscompany.bot")


class Context:
    """Message context passed to handlers."""

    def __init__(self, bot: "Bot", topic: str, from_user: str, content: str, seq_id: int = 0):
        self.bot = bot
        self.topic = topic
        self.from_user = from_user
        self.content = content
        self.seq_id = seq_id

    def reply(self, text: str):
        """Reply with text to the same topic."""
        self.bot.send_message(self.topic, text)

    def reply_with_typing(self, text: str, delay: float = 0.5):
        """Send typing indicator, wait briefly, then reply."""
        self.bot.send_typing(self.topic)
        time.sleep(delay)
        self.bot.send_message(self.topic, text)


class Bot:
    """Cats Company bot that connects via WebSocket.

    Usage:
        bot = Bot("ws://localhost:6061/v0/channels", "cc_1a_abc123...")

        @bot.on_message
        def handle(ctx):
            ctx.reply("Hello!")

        bot.run()
    """

    def __init__(self, server_url: str, api_key: str):
        self.server_url = server_url
        self.api_key = api_key
        self._ws: Optional[websocket.WebSocketApp] = None
        self._message_handler: Optional[Callable] = None
        self._ready_handler: Optional[Callable] = None
        self._running = False
        self._msg_id = 0
        self._lock = threading.Lock()

    def _next_id(self) -> str:
        with self._lock:
            self._msg_id += 1
            return str(self._msg_id)

    def on_message(self, handler: Callable) -> Callable:
        """Register a message handler. Can be used as a decorator."""
        self._message_handler = handler
        return handler

    def on_ready(self, handler: Callable) -> Callable:
        """Register a ready callback. Can be used as a decorator."""
        self._ready_handler = handler
        return handler

    def send_message(self, topic: str, text: str):
        """Send a text message to a topic."""
        msg = {"pub": {"id": self._next_id(), "topic": topic, "content": text}}
        self._send(msg)

    def send_typing(self, topic: str):
        """Send a typing indicator to a topic."""
        msg = {"note": {"topic": topic, "what": "kp"}}
        self._send(msg)

    def _send(self, msg: dict):
        """Serialize and send a message over the WebSocket."""
        if self._ws:
            try:
                self._ws.send(json.dumps(msg))
            except Exception as e:
                logger.error(f"send error: {e}")

    def _do_handshake(self, ws):
        """Send {hi} handshake and wait for ctrl 200."""
        hi = {"hi": {"id": self._next_id(), "ver": "0.1.0"}}
        ws.send(json.dumps(hi))

    def _on_open(self, ws):
        """Called when WebSocket connection is established."""
        logger.info(f"connected to {self.server_url}")
        self._do_handshake(ws)

    def _on_message(self, ws, raw: str):
        """Called for each incoming WebSocket message."""
        try:
            msg = json.loads(raw)
        except json.JSONDecodeError:
            logger.warning(f"invalid JSON: {raw[:100]}")
            return

        # Handle handshake response
        if "ctrl" in msg:
            ctrl = msg["ctrl"]
            if ctrl.get("code") == 200 and ctrl.get("params", {}).get("build") == "catscompany":
                logger.info("handshake complete")
                if self._ready_handler:
                    self._ready_handler()
            return

        # Handle data messages
        if "data" in msg:
            data = msg["data"]
            if self._message_handler:
                ctx = Context(
                    bot=self,
                    topic=data.get("topic", ""),
                    from_user=data.get("from", ""),
                    content=str(data.get("content", "")),
                    seq_id=data.get("seq", 0),
                )
                try:
                    self._message_handler(ctx)
                except Exception as e:
                    logger.error(f"handler error: {e}")

    def _on_error(self, ws, error):
        logger.error(f"websocket error: {error}")

    def _on_close(self, ws, close_status_code, close_msg):
        logger.info(f"connection closed (code={close_status_code})")
        self._running = False

    def run(self):
        """Start the bot (blocking). Reconnects on disconnect."""
        self._running = True
        url = f"{self.server_url}?api_key={self.api_key}"

        logger.info(f"bot starting, connecting to {self.server_url}")

        while self._running:
            try:
                self._ws = websocket.WebSocketApp(
                    url,
                    on_open=self._on_open,
                    on_message=self._on_message,
                    on_error=self._on_error,
                    on_close=self._on_close,
                )
                self._ws.run_forever(ping_interval=30, ping_timeout=10)
            except KeyboardInterrupt:
                logger.info("bot shutting down...")
                self._running = False
                break
            except Exception as e:
                logger.error(f"connection error: {e}")

            if self._running:
                logger.info("reconnecting in 3s...")
                time.sleep(3)

        if self._ws:
            try:
                self._ws.close()
            except Exception:
                pass
        logger.info("bot stopped")
